webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -251,38 +305,41 @@ trait CachedFindRuleNodeStatusReports extends ReportingService with CachedReposi
          *  - Other cases: no expiration
          *
          */
-        (expired, upToDate) =  nodeIds.partition { id =>  //two group: expired id, and up to date info
-                                 cache.get(id) match {
-                                   case None         => true
-                                   case Some(status) => status.runInfo match { // We have a status on the run
-                                     case t : ExpiringStatus => t.expirationDateTime.isBefore(now)
-                                     case UnexpectedVersion(_, _, lastRunExpiration, _, _)
-                                                             => lastRunExpiration.isBefore(now)
-                                     case UnexpectedNoVersion(_, _, lastRunExpiration, _, _)
-                                                             => lastRunExpiration.isBefore(now)
-                                     case _                  => false
-                                   }
-                                 }
-                               }
-        newStatus           <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(expired, Set())
+         upToDateIds =  toReturn.flatMap { case (_, report) =>  //two group: expired id, and up to date info
+                          val expired = report.runInfo match {
+                            case t : ExpiringStatus => t.expirationDateTime.isBefore(now)
+                            case UnexpectedVersion(_, _, lastRunExpiration, _, _)
+                                                    => lastRunExpiration.isBefore(now)
+                            case UnexpectedNoVersion(_, _, lastRunExpiration, _, _)
+                                                    => lastRunExpiration.isBefore(now)
+                            case _                  => false
+                          }
+                          if(expired) None else Some(report.nodeId)
+                        }
+        // in that way to not miss node ids not yet in cache
+        toUpdateIds  =  nodeIds -- upToDateIds.toSet
+        _            <- invalidate(toUpdateIds).unit.toBox
       } yield {
-        //here, newStatus.keySet == expired.keySet, so we have processed all nodeIds that should be modified.
-        ReportLogger.debug(s"Compliance cache miss (updated):[${newStatus.keySet.map(_.value).mkString(" , ")}], " +
-                           s" hit:[${upToDate.map(_.value).mkString(" , ")}]")
-        cache = cache ++ newStatus
-        val toReturn = cache.filterKeys { id => nodeIds.contains(id) }
-        ReportLogger.trace("Compliance cache content: " + cacheToLog(toReturn))
+        ReportLogger.Cache.debug(s"Compliance cache to load (expired, missing):[${toUpdateIds.map(_.value).mkString(" , ")}]")
+        ReportLogger.Cache.trace("Compliance cache hit: " + cacheToLog(toReturn))
```

Comment:
We should at least advertise loudly that the returned map may not contain all the nodes from the method parameters, but only those in cache (expired or not).